File: services/inventory_service/core/src/services/replenishment.rs
use crate::domains::replenishment::{
    CreateReorderRule, ReorderRule, ReplenishmentCheckResult, UpdateReorderRule,
};
use crate::error::AppError;
use async_trait::async_trait;
use uuid::Uuid;

/// Service trait for automated stock replenishment
#[async_trait]
pub trait ReplenishmentService: Send + Sync {
    /// Create a new reorder rule
    async fn create_reorder_rule(
        &self,
        tenant_id: Uuid,
        rule: CreateReorderRule,
    ) -> Result<ReorderRule, AppError>;

    /// Get reorder rule by ID
    async fn get_reorder_rule(
        &self,
        tenant_id: Uuid,
        rule_id: Uuid,
    ) -> Result<Option<ReorderRule>, AppError>;

    /// Update reorder rule
    async fn update_reorder_rule(
        &self,
        tenant_id: Uuid,
        rule_id: Uuid,
        updates: UpdateReorderRule,
    ) -> Result<ReorderRule, AppError>;

    /// Delete reorder rule
    async fn delete_reorder_rule(&self, tenant_id: Uuid, rule_id: Uuid) -> Result<(), AppError>;

    /// List reorder rules for a product
    async fn list_reorder_rules_for_product(
        &self,
        tenant_id: Uuid,
        product_id: Uuid,
        warehouse_id: Option<Uuid>,
    ) -> Result<Vec<ReorderRule>, AppError>;

    /// Run replenishment check for all active rules
    async fn run_replenishment_check(
        &self,
        tenant_id: Uuid,
    ) -> Result<Vec<ReplenishmentCheckResult>, AppError>;

    /// Run replenishment check for a specific product
    async fn check_product_replenishment(
        &self,
        tenant_id: Uuid,
        product_id: Uuid,
        warehouse_id: Option<Uuid>,
    ) -> Result<ReplenishmentCheckResult, AppError>;
}

File: frontend/src/lib/api/inventory/replenishment.ts
// =============================================================================
// Replenishment API Client
// Handles reorder rules and replenishment checking
// =============================================================================

import { apiClient } from '$lib/api/client';
import type { ApiResponse } from '$lib/types';
import type {
	ReorderRule,
	CreateReorderRule,
	UpdateReorderRule,
	ReplenishmentCheckResult,
	PaginationParams
} from '$lib/types/inventory';
import { buildQueryString, toRecord } from './utils';

const BASE_PATH = '/inventory/replenishment';

interface ReorderRuleListParams extends PaginationParams {
	productId?: string;
	warehouseId?: string;
}

interface ReorderRuleListResponse {
	rules: ReorderRule[];
	pagination: {
		page: number;
		pageSize: number;
		totalItems: number;
		totalPages: number;
		hasNext: boolean;
		hasPrev: boolean;
	};
}

interface ReplenishmentCheckParams {
	productId?: string;
	warehouseId?: string;
	autoTrigger?: boolean;
}

/**
 * Replenishment API client for inventory service
 */
export const replenishmentApi = {
	// =========================================================================
	// Reorder Rules
	// =========================================================================

	/**
	 * List reorder rules
	 */
	async listRules(
		params: ReorderRuleListParams = {}
	): Promise<ApiResponse<ReorderRuleListResponse>> {
		const query = buildQueryString(toRecord(params));
		return apiClient.get<ReorderRuleListResponse>(`${BASE_PATH}/rules${query}`);
	},

	/**
	 * Get a reorder rule by ID
	 */
	async getRule(ruleId: string): Promise<ApiResponse<ReorderRule>> {
		return apiClient.get<ReorderRule>(`${BASE_PATH}/rules/${ruleId}`);
	},

	/**
	 * Create a new reorder rule
	 */
	async createRule(data: CreateReorderRule): Promise<ApiResponse<ReorderRule>> {
		return apiClient.post<ReorderRule>(`${BASE_PATH}/rules`, toRecord(data));
	},

	/**
	 * Update an existing reorder rule
	 */
	async updateRule(ruleId: string, data: UpdateReorderRule): Promise<ApiResponse<ReorderRule>> {
		return apiClient.patch<ReorderRule>(`${BASE_PATH}/rules/${ruleId}`, toRecord(data));
	},

	/**
	 * Delete a reorder rule
	 */
	async deleteRule(ruleId: string): Promise<ApiResponse<void>> {
		return apiClient.delete<void>(`${BASE_PATH}/rules/${ruleId}`);
	},

	// =========================================================================
	// Replenishment Checks
	// =========================================================================

	/**
	 * Check replenishment needs for products
	 */
	async check(
		params: ReplenishmentCheckParams = {}
	): Promise<ApiResponse<ReplenishmentCheckResult[]>> {
		const query = buildQueryString(toRecord(params));
		return apiClient.get<ReplenishmentCheckResult[]>(`${BASE_PATH}/check${query}`);
	},

	/**
	 * Trigger replenishment for a specific product
	 */
	async trigger(
		productId: string,
		warehouseId?: string
	): Promise<ApiResponse<ReplenishmentCheckResult>> {
		return apiClient.post<ReplenishmentCheckResult>(`${BASE_PATH}/trigger`, {
			productId,
			warehouseId
		});
	}
};
